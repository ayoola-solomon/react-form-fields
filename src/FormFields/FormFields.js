import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

const addPropsToFields = fields => {
  const fieldsWithProps = [];

  fields.forEach(field => {
    if (field.container && field.fields) {
      const { container, fields, ...props } = field;
      fieldsWithProps.push({
        Container: container,
        fields: addPropsToFields(fields),
        ...props
      });
    } else if (field.component) {
      const { component, ...props } = field;
      fieldsWithProps.push({
        ...props,
        Component: component
      });
    }
  });

  return fieldsWithProps;
};

const FormFields = ({ fields }) => {
  const fieldsWithProps = addPropsToFields(fields);

  const renderFields = fieldsToBeRendered =>
    fieldsToBeRendered.map(field => {
      if (field.Container && field.fields) {
        const { Container, fields, ...props } = field;
        return (
          <Container key={field.name} {...props}>
            {renderFields(fields)}
          </Container>
        );
      }

      const { Component, ...props } = field;
      return (
        <Fragment key={props.name}>
          <Component {...props} />
        </Fragment>
      );
    });

  return renderFields(fieldsWithProps);
};

FormFields.propType = {
  fields: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
      component: PropTypes.element.isRequired
    })
  ).isRequired
};

export default FormFields;
