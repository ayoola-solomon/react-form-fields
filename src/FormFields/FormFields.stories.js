import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';

import FormFields from '.';
import WithFormik from '../examples/WithFormik';
import WithHooks from '../examples/WithHooks';
import { Container, TextField, FormContainer } from '../examples/styles';

import readme from '../../README.md';

const defaultFields = [
  {
    type: 'text',
    name: 'firstName',
    component: TextField,
    placeholder: 'First Name'
  },
  {
    type: 'text',
    name: 'lastName',
    component: TextField,
    placeholder: 'Last Name'
  }
];

const rowFields = [
  {
    container: Container,
    name: 'emailAndPassword',
    fields: [
      {
        ...defaultFields[0],
        placeholder: 'Email',
        type: 'email'
      },
      {
        type: 'password',
        name: 'password',
        component: TextField,
        placeholder: 'Password',
        type: 'pasword'
      }
    ]
  }
];

storiesOf('Form Fields', module)
  .addParameters({
    notes: { markdown: readme }
  })
  .addDecorator(centered)
  .add('Default', () => (
    <FormContainer>
      <h1>Default</h1>
      <FormFields fields={defaultFields} />
    </FormContainer>
  ))
  .add('In a Row', () => (
    <FormContainer>
      <h1>In a Row</h1>
      <FormFields fields={rowFields} />
    </FormContainer>
  ))
  .add('With Formik', () => <WithFormik />)
  .add('With React Hooks', () => <WithHooks />);
