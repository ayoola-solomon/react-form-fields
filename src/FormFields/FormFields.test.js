import React from 'react';
import { mount } from 'enzyme';

import FormFields from './FormFields';
import { TextField } from '../examples/styles';

const fields = [
  {
    type: 'text',
    name: 'firstName',
    component: TextField,
    placeholder: 'First Name'
  },
  {
    type: 'text',
    name: 'lastName',
    component: TextField,
    placeholder: 'Last Name'
  }
];

describe('FormFields', () => {
  it('should render properly', () => {
    const component = mount(<FormFields fields={fields} />);
    expect(component).toMatchSnapshot();
  });

  it('should throw an error fields is not available', () => {
    expect(() => mount(<FormFields />)).toThrowError(
      'You are calling FormFields with an undefined fields'
    );
  });
});
