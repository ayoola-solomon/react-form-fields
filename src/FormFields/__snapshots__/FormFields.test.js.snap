// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFields should render properly 1`] = `
<FormFields
  fields={
    Array [
      Object {
        "component": [Function],
        "name": "firstName",
        "placeholder": "First Name",
        "type": "text",
      },
      Object {
        "component": [Function],
        "name": "lastName",
        "placeholder": "Last Name",
        "type": "text",
      },
    ]
  }
>
  <TextField
    name="firstName"
    placeholder="First Name"
    type="text"
  >
    <styled.div>
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bxivhb",
              "isStatic": true,
              "lastClassName": "WdVyl",
              "rules": Array [
                "
  margin-bottom: 16px;

  p {
    margin: 0;
    color: red;
    font-size: 14px;
    margin-top: 8px;
  }
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bxivhb",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <div
          className="sc-bxivhb WdVyl"
        >
          <styled.input
            name="firstName"
            placeholder="First Name"
            type="text"
          >
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-bwzfXH",
                    "isStatic": true,
                    "lastClassName": "llWBwP",
                    "rules": Array [
                      "
  height: 30px;
  font-size: 1em;
  border: 2px solid #ddd;
  border-radius: 4px;
  width: 250px;
  padding: 8px 15px;
",
                    ],
                  },
                  "displayName": "styled.input",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-bwzfXH",
                  "target": "input",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
              name="firstName"
              placeholder="First Name"
              type="text"
            >
              <input
                className="sc-bwzfXH llWBwP"
                name="firstName"
                placeholder="First Name"
                type="text"
              />
            </StyledComponent>
          </styled.input>
        </div>
      </StyledComponent>
    </styled.div>
  </TextField>
  <TextField
    name="lastName"
    placeholder="Last Name"
    type="text"
  >
    <styled.div>
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bxivhb",
              "isStatic": true,
              "lastClassName": "WdVyl",
              "rules": Array [
                "
  margin-bottom: 16px;

  p {
    margin: 0;
    color: red;
    font-size: 14px;
    margin-top: 8px;
  }
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bxivhb",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <div
          className="sc-bxivhb WdVyl"
        >
          <styled.input
            name="lastName"
            placeholder="Last Name"
            type="text"
          >
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-bwzfXH",
                    "isStatic": true,
                    "lastClassName": "llWBwP",
                    "rules": Array [
                      "
  height: 30px;
  font-size: 1em;
  border: 2px solid #ddd;
  border-radius: 4px;
  width: 250px;
  padding: 8px 15px;
",
                    ],
                  },
                  "displayName": "styled.input",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-bwzfXH",
                  "target": "input",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
              name="lastName"
              placeholder="Last Name"
              type="text"
            >
              <input
                className="sc-bwzfXH llWBwP"
                name="lastName"
                placeholder="Last Name"
                type="text"
              />
            </StyledComponent>
          </styled.input>
        </div>
      </StyledComponent>
    </styled.div>
  </TextField>
</FormFields>
`;
